# Cleaning Used Cars
## A pandas / numpy project

The aim of this project is to clean a webscraped dataset containing info on used cars, sold on Ebay. The dataset contains 50,000 datapoints and [can be found on Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).

It must be noted that the dataset has been muddied by Dataquest.io for data cleaning practice purposes.

The cleaned dataset can then be used for analytics or machine learning prediction purposes. The aim is to use this dataset for a model predicting the price you can get for your car (separate project).

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

autos = pd.read_csv('autos.csv', encoding='latin1')

### Data exploration

Let's start with examining the dataset to get an idea of what we are working with and where to start cleaning.

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
powerPS               50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

The dataset contains 20 columns, most of which are 'non-null objects' (= strings). 

We can see that the column values in 'price' are strings. Upon closer inspection, we can see that the values contain the '$' symbol.

The 'odometer' column has text values, which need to be converted into numerical values.

We can see that there are almost 9000 cases of missing data on notRepairedDamage (showing if the car has damage which is not yet repaired).

Also, fuelType data is missing in around 4500 cases.

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model odometer_km  registration_month fuel_type  \
0    manuell      158  andere   150,000km                   3       lpg   
1  automatik      286     7er   150,000km                   6    benzin   
2    manuell      102    golf    70,000km                   7    benzin   
3  automatik       71  fortwo    70,000km                   6    benzin   
4    manuell        0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

When examining the columns and values, I suspect the model column being an open text field on Ebay. Below, I will count the number of different models.

In [7]:
autos['model'].value_counts()

golf                  4024
andere                3528
3er                   2761
polo                  1757
corsa                 1735
astra                 1454
passat                1425
a4                    1291
5er                   1183
c_klasse              1172
e_klasse              1001
a3                     882
a6                     835
focus                  811
fiesta                 769
transporter            704
twingo                 677
2_reihe                621
a_klasse               588
vectra                 574
fortwo                 574
1er                    538
clio                   509
3_reihe                508
mondeo                 496
touran                 468
punto                  458
zafira                 423
ka                     370
megane                 362
                      ... 
gl                      12
lanos                   12
range_rover_sport       12
move                    10
r19                     10
lybra                   10
m

We can see that there are 245 different models. Some model names are suspected of being the same but expressed differently (f.e. in German instead of English: 2_reihe == 2_series). We will need to transform these modelnames later on.

Next, we can see that the model 'andere' (which translates to 'other') is very popular. It is unclear what models are in the 'other' category.

In [8]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
2006    2708
2001    2703
2002    2533
1998    2453
2007    2304
2008    2231
2009    2098
1997    2028
2011    1634
2010    1597
2017    1453
1996    1444
2012    1323
2016    1316
1995    1313
2013     806
2014     666
1994     660
2018     492
1993     445
2015     399
1990     395
1992     391
1991     356
1989     181
        ... 
1950       3
1955       2
9000       2
1954       2
1800       2
1957       2
1941       2
1951       2
1934       2
4100       1
4800       1
1953       1
1111       1
1927       1
6200       1
4500       1
1943       1
5911       1
1939       1
1938       1
2800       1
8888       1
1000       1
1500       1
1948       1
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

Examining the yearOfRegistration column reveals that it contains registration dates which are unrealistic (f.e. 8888, 1000, 1500).

In [9]:
autos['price'].value_counts().head()

$0        1421
$500       781
$1,500     734
$2,500     643
$1,000     639
Name: price, dtype: int64

As we can see, we need to remove the dollar signs and turn the values into numerical values so we can analyse prices further.

We can see that there are many cars without price in the list, scewing the distribution. It is possible that these cars are listed without price so that people can bid on them

It would be interesting to split cars with and without price.

In [10]:
autos.describe()

registration_year       powerPS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

Going through the columns, we can easily spot some unrealistic values:

Year of Registration contains a max value of 9999 and a min of 1000.

powerPS (metric horsepower) has a min value of 0 and a max value 17700.

Also, every car in the dataset has 0 number of pictures.

Month of registration min is 0. This should probably be a null-value instead.

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Cleaning

From the table above, we can note that some columns contain mostly one value (and are thus not adding anything to our analysis):

- seller: This column contains only 1 'commercial' value, the rest is 'privat'.
- offer_type
- nr of pictures: This column contains only '0.0' values.

We will drop these columns.

Another thing striking is the top vehicle type: Limousine.
- Investigate vehicle types.

As seen earlier, we will perform the following cleaning actions:

- Remove $ from 'price' and turn it into integer.
- Turn 'odometer' into integer.
- Split price between '0' and '>0'
- Remove unrealistic prices, registration years, powerPS, odometer, registration month.
- Investigate 'unrepaired damage' null-values.
- Investigate 'name' and 'model'.

In [12]:
## dropping columns
autos = autos.drop(columns = ['seller', 'offer_type', 'nr_of_pictures'])

## removing any non-numeric character from price and odometer
autos['price'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
autos['odometer_km'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')

## changing datatype into integer
autos['price'] = autos['price'].astype('int64');
autos['odometer_km'] = autos['odometer_km'].astype('int64');

In [13]:
autos['price'].head(3)

0    5000
1    8500
2    8990
Name: price, dtype: int64

In [14]:
autos['odometer_km'].head(3)

0    150000
1    150000
2     70000
Name: odometer_km, dtype: int64

In [15]:
## determining threshold for price
autos['price'].sort_values().tail(10)

7814      1300000
47634     3890000
11137    10000000
24384    11111111
2897     11111111
27371    12345678
47598    12345678
39377    12345678
42221    27322222
39705    99999999
Name: price, dtype: int64

We can see that the prices are becoming unrealistic from 10 million and upwards.

In [16]:
autos.sort_values(by='price')

date_crawled                                               name  \
18089  2016-03-16 15:51:44                     VW_Polo_preiswert_zu_verkaufen   
43923  2016-03-25 00:56:44                     Focus_Diesel_zum_ausschlachten   
15225  2016-03-12 10:55:13                                       Audi_a3_1_8T   
29499  2016-03-24 21:56:38                     _TUNING_GRUNE_PLAKETTE_EURO_4_   
8445   2016-03-25 20:53:17                                    Suche_Audi_a6_4   
38832  2016-03-10 15:47:28  Opel_Corsa_B_mit_Schiebedach_Alufelgen_&_TÜV_b...   
43925  2016-03-15 12:51:38                       Golf_3_gti_Jubilaeum_Edition   
8438   2016-03-14 21:37:21         Jaguar_xj_40_Daimler_Original_Baujahr_1994   
35821  2016-04-01 18:48:07  Tausche_gegen_Gold_/_Silber_oder_gleichwertige...   
31332  2016-03-07 15:56:14                           Renault_clio_1.2_16_v_RT   
12819  2016-03-08 21:49:42                                   VW_Bora__Tausch_   
33620  2016-03-26 18:39:00  VW_Jetta_Oldtimer_Youngtimer_Sammler_Retro_1.8...   
15208  2016-03-19 09:39:15      BMW_E39_520_Tempomat_Xenon_TÜV_4/17_188.000km   
2481   2016-04-04 11:47:53  Triton_Supermoto_400_ccm_Tuev_9/17_!_Verkauf_/...   
4266   2016-03-08 17:53:09                                    Passat_3b_kombi   
27020  2016-03-31 16:38:27  wollen_Sie_Ihr_Auto_Verkaufen___wir_Kaufen_KFZ_an   
35819  2016-03-24 10:56:50                                  Alfa_Romeo_Spider   
31336  2016-03-21 08:36:17  Peugeot_206_1_4_75_PS_TOP_im_Tausch_gegen_Komb...   
29482  2016-04-03 20:58:55                                     Daihatsu_Cuore   
47861  2016-03-29 11:54:52                           W211_Stossdaempfer_vorne   
21253  2016-03-12 15:49:05                               Audi_80_B4_Limousine   
37021  2016-03-08 17:39:51                         Suche_SLK__R_170_Automatik   
29503  2016-03-09 10:47:37                         Golf_3___Baujahr_96__1_6_l   
21256  2016-03-20 11:36:56                  VW_GOLF_2_TÜV_BBS_Tausch_moeglich   
2470   2016-03-05 15:49:58                                   Opel_Astra_Coupe   
24871  2016-03-29 09:37:36    Golf_6_GTD_DSG_*Standheizung*_*Tausch_moeglich*   
31264  2016-04-02 22:38:54                                       opel_astra_g   
1220   2016-03-30 13:55:32  Opel_Corsa_C_1__2_Liter_mit_Klima_TÜV_bis_10.2017   
24900  2016-03-16 14:37:54                                 VW_Golf_4___Diesel   
21171  2016-03-11 19:50:48                        Opel_vectra_b_super_Zustand   
...                    ...                                                ...   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
514    2016-03-17 09:53:08                   

When examining the most expensive cars, we can see that it doesn't make sense. Exclusive ferrari's second hand on Ebay along with expensive Volkswagen busses. We remove these values.

In [17]:
## removing outliers price
autos = autos[(autos["price"] <= 350000)]

Another striking thing is the high number (1421) of cars with price '0'. Let's examine this category further:

In [18]:
autos[autos.price == 0]

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

A few observations when examining price == 0 cars:
- In the name column we can see that some listings are 'suche' (= 'looking for'). This should be a separate category.
- It could be that people can bid on cars which are listed without price (or a very low price (1), see below).

For now, we are only interested in used cars being sold.

In [19]:
## showing cars with low price
autos[autos.price <= 10]

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
344    2016-03-22 11:25:18  Verkaufe_hier_mein_mein_schoener_honda_crx_in_...   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
...                    ...                                                ...   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49240  2016-03-10 09:54:48   Suche_7_Sitzer_Espace_oder?_Biete_s.Beschreibung   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49377  2016-03-20 00:58:36             Tausche_BMW_318i_Facelift_Model_143_ps   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e36_318_i_Limo_Ringtool_PS_Tuning   
49528  2016-03-28 19:57:41       BMW_E36__316i_Compact_M_Paket_"Schlachtfest"   
49538  2016-03-14 10:55:27                 Golf_3_mit_Tuev_.....PREISWERT!!!!   
49551  2016-03-16 23:40:18        Golf_IV_an_

In [20]:
## splitting off 'suche' category 
suche = autos[autos['name'].str.contains("suche|Suche|SUCHE")]
## splitting off price=0 cars into separate price_0 df
price_0 = autos[autos.price <= 10]
## removing price=0 and suche cars from autos df
autos = autos[autos.price > 10]
autos = autos[~autos['name'].str.contains("suche|Suche|SUCHE")]

In [21]:
suche.describe()

price  registration_year     powerPS    odometer_km  \
count     246.000000         246.000000  246.000000     246.000000   
mean     2206.463415        1997.398374   28.654472  104329.268293   
std      9109.847358          34.469564   65.906230   61392.492038   
min         0.000000        1500.000000    0.000000    5000.000000   
25%         0.000000        1998.250000    0.000000   20000.000000   
50%        18.000000        2000.000000    0.000000  150000.000000   
75%      1875.000000        2005.000000    0.000000  150000.000000   
max    130000.000000        2019.000000  450.000000  150000.000000   

       registration_month   postal_code  
count          246.000000    246.000000  
mean             1.414634  43801.101626  
std              2.974523  25346.041383  
min              0.000000   1979.000000  
25%              0.000000  26127.000000  
50%              0.000000  40719.000000  
75%              1.000000  59348.000000  
max             12.000000  99974.000000

In [22]:
## determining odometer outliers threshold
autos['odometer_km'].value_counts()

150000    31224
125000     5037
100000     2099
90000      1731
80000      1414
70000      1215
60000      1154
50000      1010
40000       814
5000        791
30000       778
20000       756
10000       247
Name: odometer_km, dtype: int64

We can see that the number of car with 5000km doesn't really fit the descending count pattern. An explanation for this is that 5000 is the lowest category to choose from, so it includes 5000km and below.

In [23]:
## examining powerPS
autos['powerPS'].value_counts().head()

0      4817
75     3056
60     2129
150    1963
140    1862
Name: powerPS, dtype: int64

In [24]:
autos['powerPS'].sort_values().tail()

16743    14009
22592    15001
46986    16011
35039    16312
36421    17700
Name: powerPS, dtype: int64

We can see that the dataset contains cars with unrealistic horsepower values. The strongest car in the world has around 1300 horsepower, so we should drop anything above that.
Also, cars with 0 horsepower don't exist.

In [25]:
## dropping powerPS values above 1300
autos = autos[autos.powerPS <= 1300]
## dropping powerPS under 5 values 
autos = autos[autos.powerPS >= 5]

In [26]:
## registration years
autos['registration_year'].sort_values().tail()

49185    2019
27578    2800
22799    5000
27618    5911
49910    9000
Name: registration_year, dtype: int64

We can see that there are cars registered in the future in this dataset. We will remove these. 

In [27]:
autos.sort_values(by='registration_year', ascending=False)

date_crawled                                               name  \
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   
27618  2016-03-29 22:39:32  Golf_1_75ps_5911km_Original_Automatik_einer_de...   
22799  2016-03-20 18:56:44                                  Subaru_Impreza_GT   
27578  2016-03-31 22:49:04                                   VW_GOLF_2_SYNCRO   
49185  2016-03-28 11:45:31              Alfa_Romeo_156_Sportwagon_1_8_T_Spark   
38617  2016-03-26 18:58:00  VW_Beetel_TÜV_NEU__sehr_gepflegt_und_technisch...   
40254  2016-03-17 17:54:53                                Mercedes_Benz_E_240   
41266  2016-03-23 16:50:38                                         fiat_punto   
43901  2016-03-27 15:52:30                                      Polo_6n2_1.4l   
35047  2016-03-29 14:25:27  peugeot_206_tuev_neu_04.2018_mit_magnaflow_spo...   
31664  2016-03-23 15:47:31                                  MERCEDES_GL350CDI   
13019  2016-03-23 16:47:32  A_140___1._Hd__nur_97tkm__TÜV_neu__Sch_Heft__u...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
49218  2016-04-01 19:48:06          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev   
8882   2016-03-15 15:54:39   Volvo_C30_1.6D_130294km__TÜV_bis_02.2018___Leder   
22969  2016-03-23 18:51:17                Mercedes_A160_Avangarde_**TÜV_Neu**   
40375  2016-04-04 17:50:07                                   Opel_Corsa_D_OPC   
23314  2016-03-26 22:38:19  Gepflegter_VW_Touran_1.9TDI_Standheizung_Klima...   
18345  2016-03-11 09:50:36      Mitsubischi_colt_zc3_tuev_2/18_anfaenger_auto   
15475  2016-03-29 16:56:14                         Ford_Galaxy_1.9_TDI_Top!!!   
20620  2016-04-05 01:50:30  ___Opel_Astra_Kombi_1.7_CDTI___TÜV_bis_01.2018___   
11239  2016-03-12 15:45:30                     Ford_ka_mit_neu_tuev_und_Klima   
45825  2016-03-20 16:56:24                               Smart__mit_ATM_Motor   
35021  2016-03-15 00:53:06                                      Vw_Passat_1.6   
13835  2016-03-24 10:56:32          Mercedes_E_200_Automatik_74060km_Tuev_neu   
19626  2016-03-12 13:37:40                            fahrbereiter_Seat_Ibiza   
33580  2016-03-25 10:50:40  BMW_e46_320td_Compact__8_fach_bereift__Tuev_bi...   
26542  2016-03-08 21:54:46         Ford_Focus_Style__TÜV_NEU__Klima__Tempomat   
2785   2016-03-30 12:55:06                                          Seat_Leon   
33738  2016-04-02 15:36:21                     Gepflegter_Seat_Ibiza_in_weiss   
...                    ...                                                ...   
7094   2016-03-08 16:45:15  Borgward_isabella_restauration_Guenstiger_Oldt...   
1548   2016-03-25 14:46:44               1936_Daimler_Fifteen_Special_Projekt   
36822  2016-03-16 08:56:42                                      Triumph_TR3_A   
24291  2016-03-05 14:45:08                                         Ford_F_100   
9620   2016-03-05 21:53:43                       Sehr_schoener_Plymouth_Savoy   
13366  2016-03-26 12:40:12                               Andere_Panhard_PL_17   
20471  2016-03-28 13:38:56      Biete_312_Wartburg_Limousine__Baujahr_05/1959   
23854  2016-03-28 12:59:22                                      Ford_Fairlane   
34541  2016-04-02 01:57:41            Mercedes_Benz_180_D_Ponton_mit_Faltdach   
10358  2016-03-19 13:45:25                          Wartburg_311_zu_verkaufen   
20105  2016-03-18 21:37:22  VW_Winker_Faltdach_Kaefer_Bj_1958_zum_Fertigst...   
23044  2016-03-20 15:54:00  Mercedes_Benz_Ponton_220s_mit_grossem_Faltdach...   
40077  2016-03-12 23:57:10  Andere_Horch_Sachsenring_P2M_Kuebelwagen_DDR_N...   
7294   2016-03-28 16:26:34      YA126_DAF_Berge__Werstatt_und_Manschaft_Wagen   
26570  2016-03-20 16:52:36                                Opel_Olympia_Rekord   
19914  2016-03-06 10:51:30  Massey_Ferguson_MF35_FE35_Traktor_Trecker_Dies...   
36794  2016-04-01 09:47:52  Mercedes_Benz_190SL_FIN_Nr._63_top_History__In...   
25556  2016-03-29 11:47:24            Chevrol

In [28]:
autos['registration_year'].sort_values().head()

21416    1927
22101    1929
11246    1931
21421    1937
26607    1937
Name: registration_year, dtype: int64

The oldest car in the set is from 1927, which on closer inspection turns out to be an oldtimer. We will keep these values.

Since the set was scraped in 2016, we will carefully inspect values between 2016 and 2019 (it could be that DataQuest added these observations to muddy the set).

In [29]:
autos.sort_values(by='registration_year', ascending=False)

date_crawled                                               name  \
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   
27618  2016-03-29 22:39:32  Golf_1_75ps_5911km_Original_Automatik_einer_de...   
22799  2016-03-20 18:56:44                                  Subaru_Impreza_GT   
27578  2016-03-31 22:49:04                                   VW_GOLF_2_SYNCRO   
49185  2016-03-28 11:45:31              Alfa_Romeo_156_Sportwagon_1_8_T_Spark   
38617  2016-03-26 18:58:00  VW_Beetel_TÜV_NEU__sehr_gepflegt_und_technisch...   
40254  2016-03-17 17:54:53                                Mercedes_Benz_E_240   
41266  2016-03-23 16:50:38                                         fiat_punto   
43901  2016-03-27 15:52:30                                      Polo_6n2_1.4l   
35047  2016-03-29 14:25:27  peugeot_206_tuev_neu_04.2018_mit_magnaflow_spo...   
31664  2016-03-23 15:47:31                                  MERCEDES_GL350CDI   
13019  2016-03-23 16:47:32  A_140___1._Hd__nur_97tkm__TÜV_neu__Sch_Heft__u...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
49218  2016-04-01 19:48:06          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev   
8882   2016-03-15 15:54:39   Volvo_C30_1.6D_130294km__TÜV_bis_02.2018___Leder   
22969  2016-03-23 18:51:17                Mercedes_A160_Avangarde_**TÜV_Neu**   
40375  2016-04-04 17:50:07                                   Opel_Corsa_D_OPC   
23314  2016-03-26 22:38:19  Gepflegter_VW_Touran_1.9TDI_Standheizung_Klima...   
18345  2016-03-11 09:50:36      Mitsubischi_colt_zc3_tuev_2/18_anfaenger_auto   
15475  2016-03-29 16:56:14                         Ford_Galaxy_1.9_TDI_Top!!!   
20620  2016-04-05 01:50:30  ___Opel_Astra_Kombi_1.7_CDTI___TÜV_bis_01.2018___   
11239  2016-03-12 15:45:30                     Ford_ka_mit_neu_tuev_und_Klima   
45825  2016-03-20 16:56:24                               Smart__mit_ATM_Motor   
35021  2016-03-15 00:53:06                                      Vw_Passat_1.6   
13835  2016-03-24 10:56:32          Mercedes_E_200_Automatik_74060km_Tuev_neu   
19626  2016-03-12 13:37:40                            fahrbereiter_Seat_Ibiza   
33580  2016-03-25 10:50:40  BMW_e46_320td_Compact__8_fach_bereift__Tuev_bi...   
26542  2016-03-08 21:54:46         Ford_Focus_Style__TÜV_NEU__Klima__Tempomat   
2785   2016-03-30 12:55:06                                          Seat_Leon   
33738  2016-04-02 15:36:21                     Gepflegter_Seat_Ibiza_in_weiss   
...                    ...                                                ...   
7094   2016-03-08 16:45:15  Borgward_isabella_restauration_Guenstiger_Oldt...   
1548   2016-03-25 14:46:44               1936_Daimler_Fifteen_Special_Projekt   
36822  2016-03-16 08:56:42                                      Triumph_TR3_A   
24291  2016-03-05 14:45:08                                         Ford_F_100   
9620   2016-03-05 21:53:43                       Sehr_schoener_Plymouth_Savoy   
13366  2016-03-26 12:40:12                               Andere_Panhard_PL_17   
20471  2016-03-28 13:38:56      Biete_312_Wartburg_Limousine__Baujahr_05/1959   
23854  2016-03-28 12:59:22                                      Ford_Fairlane   
34541  2016-04-02 01:57:41            Mercedes_Benz_180_D_Ponton_mit_Faltdach   
10358  2016-03-19 13:45:25                          Wartburg_311_zu_verkaufen   
20105  2016-03-18 21:37:22  VW_Winker_Faltdach_Kaefer_Bj_1958_zum_Fertigst...   
23044  2016-03-20 15:54:00  Mercedes_Benz_Ponton_220s_mit_grossem_Faltdach...   
40077  2016-03-12 23:57:10  Andere_Horch_Sachsenring_P2M_Kuebelwagen_DDR_N...   
7294   2016-03-28 16:26:34      YA126_DAF_Berge__Werstatt_und_Manschaft_Wagen   
26570  2016-03-20 16:52:36                                Opel_Olympia_Rekord   
19914  2016-03-06 10:51:30  Massey_Ferguson_MF35_FE35_Traktor_Trecker_Dies...   
36794  2016-04-01 09:47:52  Mercedes_Benz_190SL_FIN_Nr._63_top_History__In...   
25556  2016-03-29 11:47:24            Chevrol

We can see that the cars with registration year 2018 have their ad created in 2016, which means that the registration years are wrong. The rest of the values for these cars seem legit, so we could either change these values into NaN (missing / null) values, or drop them.

Since it is only about 1500 observations, we drop these as well.

In [31]:
autos = autos[autos.registration_year < 2017]

When examining 'registration_month' we can see that there are a lot of '0' values. We could drop these, but it would mean a high loss of data (almost 2500 observations). Since we are not planning on using 'registration month' for analysis, let's not drop these values. We can interpret '0' here as 'don't know which month in this year'.

In [39]:
autos['registration_month'].value_counts()

3     4495
6     3800
4     3640
5     3628
7     3438
10    3269
9     3021
11    2977
12    2975
1     2867
8     2782
2     2663
0     2457
Name: registration_month, dtype: int64

In [42]:
## investigate model
autos['model'].value_counts()

golf                  3407
andere                2979
3er                   2433
polo                  1420
corsa                 1396
passat                1237
astra                 1226
a4                    1156
5er                   1068
c_klasse              1055
e_klasse               874
a3                     759
a6                     751
focus                  698
fiesta                 635
transporter            611
2_reihe                559
twingo                 509
1er                    507
a_klasse               489
fortwo                 488
vectra                 472
mondeo                 444
3_reihe                432
clio                   416
touran                 388
punto                  368
zafira                 348
ibiza                  304
octavia                292
                      ... 
range_rover_sport       10
mii                     10
300c                    10
sirion                  10
lanos                    9
gl                       9
r

Observations:
- 'andere' (translates to 'other') is the second largest category in models. We could examine what kind of cars these are (see below). It turns out to be a mix of rare cars and regular ones. It will be hard to split this category.
- A lot of the models have similar namings but different: range_rover, range_rover_evoque. 3_reihe, 3er. Etcetera. It will be hard to group these models.

In [48]:
autos[autos.model == 'andere']

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
45     2016-03-12 17:39:55                                 Toyota_Starlet_1.3   
70     2016-03-16 13:49:00                            Saab_9_3_2.0i_Cabrio_SE   
75     2016-04-02 18:52:03                 Volvo_960/964_Limousine_6_Zylinder   
95     2016-03-30 15:59:17                         Jaguar_XJ_5.0_V8_Portfolio   
97     2016-04-01 16:54:40                         Treuer_Begleiter_abzugeben   
136    2016-04-03 18:57:17                    2_mal_Subaru_Libero_super_delux   
149    2016-03-12 13:51:26                                     Kia_Sephia_GTX   
153    2016-03-24 10:52:19  Porsche_Porsche_968___Turbo_S_Optik__Speed_Lin...   
162    2016-03-28 10:54:23                    Fiat_Grande_Punto_1.4_16V_Sport   
163    2016-03-07 17:50:53                              Citroën_Xsara_1.8i_SX   
166    2016-04-03 20:50:55                                Mercedes_Benz_200/8   
173    2016-03-14 14:47:01                            Audi_S4_Avant_tiptronic   
195    2016-04-03 11:43:27                    Fiat_Grande_Punto_1.4_16V_Sport   
200    2016-03-28 20:41:51  Corvette_C6_Grand_Sport_Cabrio_Schalter_EU_Mod...   
201    2016-03-20 10:50:58  Mazda_2_Klima_Tuev__bis_12/_16_4_El._Fensterheber   
217    2016-03-25 14:54:53                              PORSCHE_924S__11/1989   
241    2016-03-27 21:56:01                            Hyundai_Accent_1.3i_GLS   
246    2016-03-11 06:37:33                                  Chevrolet_Pick_up   
249    2016-03-18 17:37:34                   Dodge_Caliber_SXT_1.8_Benzin_Gas   
275    2016-03-07 18:41:56     Chrsyler_Neon_gepflegt_wegen_Neukauf_abzugeben   
313    2016-03-06 22:41:57                          Kia_Clarus_1_8_SLX_116_PS   
408    2016-03-15 21:49:23                      Mitsubishi_Space_Star_mit_AHK   
446    2016-03-16 09:47:04                             Suzuki_Baleno_Tuev_Neu   
494    2016-04-02 14:45:48                                 Mazda_6__2.0Diesel   
500    2016-03-08 16:38:34                     Schicker_Cadillac_Escalade_EXT   
518    2016-03-28 16:38:38                            Jaguar_XK_3.5_Coupe_TOP   
531    2016-04-04 16:55:12                                      Suzuki_Vitara   
537    2016-03-22 14:25:25                   Audi_Cabriolet_1.8___TOP_ZUSTAND   
...                    ...                                                ...   
49376  2016-04-01 09:45:43                              Fiat_Barchetta_Cabrio   
49396  2016-03-15 19:37:20                                  Mercedes_Benz_240   
49554  2016-03-19 14:45:42  Citroën_Xantia_Kombi_2.0_HDi_Millesime__AU.neu...   
49569  2016-04-07 13:36:20                               Mitsubishi_L_300_KAT   
49586  2016-03-07 01:37:01                                  Toyota_Carina_1.6   
49621  2016-03-09 23:55:28                      Chevrolet_Nubira_1.6_Kombi_SX   
49624  2016-03-17 14:37:15  Mercedes_Benz_W126_Coupe__500_SEC_H_Zulassungs...   
49633  2016-04-03 22:36:22                 Renault_Wind_TCe_100_Night_and_Day   
49665  2016-03-21 00:58:38                        Opel_Frontera_Sport_Mistral   
49667  2016-03-07 16:53:16                            Mercedes_Benz_300_CE_24   
49680  2016-04-03 11:47:08  Volvo_Kombi_GL_940_mit_tuev_und_evtl_einen_Hae...   
49684  2016-03-26 15:53:54  Peugeot_807_HDi_130_Premium_2_2Liter_Navi_AHK_...   
49713  2016-03-22 08:51:10                         Toyota_Carina_1600_GL_TA60   
49716  2016-03-13 17:36:57                                     Peugeot_Expert   
49761  2016-03-17 19:41:33           Mercedes_W124__Youngtimer_bald_Oldtimer_   
49764  2016-03-07 21:52:19  Toyota_Land_Cruiser_Jagt_Forst_Offroad_Winde_2...   
49791  2016-03-24 15:52:51                           Toyota_IQ_1.0_Multidrive   
49805  2016-03-16 20:40:05                   

Let's move on to the date values. They are currently string values and we need to convert them in order to analyse them further:

- date_crawled
- ad_created
- last_seen

These values have both date and time values in one string. We can examine the dates by selecting only the date characters (first 10).

In [55]:
## distribution of dates in percentages
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025374
2016-03-06    0.014282
2016-03-07    0.036394
2016-03-08    0.033348
2016-03-09    0.032895
2016-03-10    0.032300
2016-03-11    0.032110
2016-03-12    0.037180
2016-03-13    0.015829
2016-03-14    0.036656
2016-03-15    0.034157
2016-03-16    0.029206
2016-03-17    0.031467
2016-03-18    0.012830
2016-03-19    0.034419
2016-03-20    0.038513
2016-03-21    0.037180
2016-03-22    0.032538
2016-03-23    0.031658
2016-03-24    0.029492
2016-03-25    0.031729
2016-03-26    0.032348
2016-03-27    0.030706
2016-03-28    0.035133
2016-03-29    0.033800
2016-03-30    0.033705
2016-03-31    0.031872
2016-04-01    0.034014
2016-04-02    0.035609
2016-04-03    0.038989
2016-04-04    0.036823
2016-04-05    0.012996
2016-04-06    0.003071
2016-04-07    0.001381
Name: date_crawled, dtype: float64

The data has been crawled between March and April 2016.

In [117]:
autos['registration_year'].describe()

count    42012.000000
mean      2003.168857
std          6.836022
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [120]:
autos['unrepaired_damage'].value_counts(dropna=False)

nein    32091
NaN      5923
ja       3998
Name: unrepaired_damage, dtype: int64

The 'unrepaired_damage' column is a strange one. It includes more missing values than 'ja' (= yes) values. My hypothesis is that sellers rather not answer this question if they have unrepaired damage to the car, thus a large portion of the missing values should be 'ja'. However, we can't tell right now and dropping all the missing values would lead to 6000 datapoints being lost, which is too large.

### Aggregation

We can continue with aggregation by brand for analysis purposes.

In [60]:
autos['brand'].nunique()

40

There are 40 unique brands in the list. Below is the top 5.

In [90]:
autos['brand'].value_counts().head()

volkswagen       8912
bmw              4789
opel             4437
mercedes_benz    4058
audi             3758
Name: brand, dtype: int64

In [100]:
autos.groupby([autos.brand == 'volkswagen']).describe()

price                                                          \
         count         mean          std   min     25%     50%     75%   
brand                                                                    
False  33100.0  6479.069456  9946.099028  11.0  1400.0  3500.0  7999.0   
True    8912.0  5704.942662  6352.681745  30.0  1300.0  3300.0  7990.0   

                registration_year                ...    registration_month  \
            max             count         mean   ...                   75%   
brand                                            ...                         
False  350000.0           33100.0  2003.299698   ...                   9.0   
True    64500.0            8912.0  2002.682899   ...                   9.0   

            postal_code                                                        \
        max       count          mean           std     min      25%      50%   
brand                                                                           
False  12.0     33100.0  52086.994532  25634.195355  1067.0  31787.0  51379.0   
True   12.0      8912.0  49911.689071  26032.807974  1067.0  28863.0  47877.0   

                          
            75%      max  
brand                     
False  72622.00  99998.0  
True   71680.75  99998.0  

[2 rows x 48 columns]

Our quick analysis above shows some statistics for volkswagen cars compared to the rest of the brands:

- 8912 of the roughly 40000 cars are volkswagen
- The mean price is 5700 euro, which is lower than the mean price of other brands combined.

Next, we want to compare the volkswagen brand to other brands individually. Below we can find the mean values for the numerical columns for each brand aggregated.

In [116]:
autos.groupby([autos.brand]).mean().round(0).sort_values(by='price', ascending=False)

price  registration_year  powerPS  odometer_km  \
brand                                                              
porsche         48705.0             2002.0    309.0      97814.0   
land_rover      19666.0             2004.0    157.0     118077.0   
sonstige_autos  14267.0             1995.0    183.0      94985.0   
jaguar          12195.0             2002.0    241.0     127077.0   
jeep            12037.0             2003.0    191.0     127806.0   
mini            10677.0             2008.0    132.0      88392.0   
audi             9742.0             2004.0    168.0     128604.0   
mercedes_benz    9011.0             2002.0    167.0     130678.0   
bmw              8617.0             2003.0    175.0     132509.0   
chevrolet        6697.0             2004.0    144.0      97489.0   
skoda            6651.0             2007.0    108.0     110097.0   
kia              6218.0             2007.0    122.0     112076.0   
dacia            6143.0             2011.0     89.0      82061.0   
volkswagen       5705.0             2003.0    107.0     128267.0   
hyundai          5605.0             2007.0    110.0     104733.0   
toyota           5298.0             2005.0    110.0     115801.0   
volvo            5177.0             2001.0    154.0     138397.0   
nissan           5142.0             2004.0    110.0     116968.0   
seat             4696.0             2005.0    101.0     120566.0   
mazda            4347.0             2003.0    119.0     124281.0   
honda            4300.0             2001.0    116.0     121701.0   
suzuki           4292.0             2004.0     89.0     108414.0   
alfa_romeo       4170.0             2002.0    146.0     131813.0   
subaru           4121.0             2000.0    143.0     125815.0   
citroen          3935.0             2005.0     97.0     118929.0   
ford             3925.0             2003.0    108.0     124258.0   
smart            3768.0             2006.0     63.0      98274.0   
chrysler         3632.0             2001.0    156.0     133469.0   
lancia           3590.0             2000.0    103.0     124419.0   
mitsubishi       3551.0             2002.0    115.0     125698.0   
saab             3376.0             2000.0    160.0     144028.0   
peugeot          3258.0             2004.0    101.0     126261.0   
opel             3162.0             2002.0    101.0     129194.0   
lada             3021.0             2000.0     76.0      85952.0   
fiat             2967.0             2004.0     80.0     116823.0   
renault          2661.0             2003.0     91.0     126958.0   
trabant          2014.0             1981.0     28.0      54875.0   
rover            1778.0             2000.0    112.0     136200.0   
daihatsu         1751.0             2003.0     68.0     117263.0   
daewoo           1072.0             2002.0     80.0     123962.0   

                registration_month  postal_code  
brand                                            
porsche                        6.0      57755.0  
land_rover                     6.0      54139.0  
sonstige_autos                 6.0      50771.0  
jaguar                         6.0      53726.0  
jeep                           7.0      54843.0  
mini                           6.0      56452.0  
audi                           6.0      55336.0  
mercedes_benz                  6.0      52013.0  
bmw                            6.0      55442.0  
chevrolet                      6.0      48483.0  
skoda                          6.0      49022.0  
kia                            6.0      49865.0  
dacia                          6.0      50770.0  
volkswagen                     6.0      49912.0  
hyundai                        6.0      49703.0  
toyota                         6.0      53053.0  
volvo                          6.0      47727.0  
nissan                         6.0      46972.0  
seat                           6.0      52472.0  
mazda                          6.0      50443.0  
honda                          6.0      47338.0  
suzuki     

## Concluding remarks and possible next steps

This concludes this data cleaning project. This cleaned dataset can now be used for machine learning purposes, for example to predict the price for which you can sell your car, or to see if a car is listed on Ebay below it's market value.

Further calculations could be done if we could scrape data on when a car was sold (listing is removed from Ebay). With that data we could measure how long it takes for a car to be sold.

Another next step would be a deepdive into the damage column. Currently, it has a lot of missing values. It would be interesting to build a machine learning model to predict the values of 'unrepaired_damage', using the values we do have in this set.